File: samples/src/main/java/io/github/danielbunting/clickhouse/samples/Pooling.java
package io.github.danielbunting.clickhouse.samples;

import io.github.danielbunting.clickhouse.pool.ClickHouseConnectionPool;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.Future;

/**
 * Concurrency with the connection pool — the JVM answer to CH.Native's
 * {@code Samples.Hosting} (app-lifetime integration).
 *
 * <p>A single {@code ClickHouseConnection} is one socket and a stateful protocol:
 * it must be used by one thread at a time (concurrent use fails fast with
 * {@code ConcurrentConnectionUseException} rather than corrupting the stream).
 * The unit of concurrency is therefore the {@link ClickHouseConnectionPool}:
 * {@code N} independent connections, one borrower each.
 *
 * <p>Steps performed:
 * <ol>
 *   <li>Create a pool of 4 connections (eagerly opened — fails fast if the
 *       server is unreachable).</li>
 *   <li>Run 16 tasks on 8 threads, each borrowing via {@code withConnection}
 *       (borrow → work → return, leak-proof).</li>
 *   <li>Show pool health: size vs. available before/during/after.</li>
 * </ol>
 *
 * <p>Hygiene the pool gives you for free: returned connections that were
 * poisoned (broken socket, abandoned mid-INSERT) are discarded, not recycled;
 * {@code validateOnBorrow} (default on) additionally health-checks idle
 * connections with {@code SELECT 1}; and discarded slots self-heal — the next
 * borrow opens a fresh connection.
 *
 * <p>Run with:
 * <pre>{@code
 *   ./gradlew :samples:runPooling
 * }</pre>
 */
public final class Pooling {

    /** The table targeted by this demo. */
    private static final String TABLE = "pool_demo";

    /** Pool capacity used by the demo. */
    private static final int POOL_SIZE = 4;

    /** Number of concurrent tasks thrown at the pool. */
    private static final int TASKS = 16;

    /** Private constructor — utility-style entry point. */
    private Pooling() {}

    /**
     * Entry point. Builds the pool and runs concurrent inserts and reads through it.
     *
     * @param args ignored
     * @throws Exception if a worker task fails
     */
    public static void main(String[] args) throws Exception {
        System.out.println("=== clickhouse-native-client Pooling tour ===");

        // --------------------------------------------------------------------
        // 1. Pool construction: 4 eagerly-opened, independent connections
        // --------------------------------------------------------------------
        System.out.println("\n[1] Opening a pool of " + POOL_SIZE + " connections...");
        try (ClickHouseConnectionPool pool =
                     ClickHouseConnectionPool.create(ClickHouseEnv.config(), POOL_SIZE)) {
            System.out.println("    size=" + pool.size() + ", available=" + pool.available());

            pool.useConnection(conn -> conn.execute(
                    "CREATE OR REPLACE TABLE " + TABLE + " ("
                            + "task UInt32,"
                            + "n    UInt64"
                            + ") ENGINE = MergeTree ORDER BY (task, n)"));

            // ----------------------------------------------------------------
            // 2. 16 tasks on 8 threads, 4 connections: borrows queue fairly
            // ----------------------------------------------------------------
            System.out.println("\n[2] Running " + TASKS + " concurrent tasks on 8 threads...");
            ExecutorService executor = Executors.newFixedThreadPool(8);
            try {
                List<Future<Long>> futures = new ArrayList<>(TASKS);
                for (int t = 0; t < TASKS; t++) {
                    final int task = t;
                    futures.add(executor.submit(() ->
                            // withConnection = borrow -> work -> always return.
                            // Each lambda runs on its own borrowed connection;
                            // no two tasks ever share a socket.
                            pool.withConnection(conn -> {
                                conn.execute("INSERT INTO " + TABLE
                                        + " SELECT " + task + ", number FROM numbers(10000)");
                                return conn.executeScalar(
                                        "SELECT count() FROM " + TABLE
                                                + " WHERE task = " + task);
                            })));
                }
                long totalSeen = 0;
                for (Future<Long> f : futures) {
                    totalSeen += f.get();
                }
                System.out.println("    All tasks done; per-task counts sum to " + totalSeen);
            } finally {
                executor.shutdown();
            }

            // ----------------------------------------------------------------
            // 3. Pool state after the burst
            // ----------------------------------------------------------------
            long rows = pool.withConnection(conn ->
                    conn.executeScalar("SELECT count() FROM " + TABLE));
            System.out.println("\n[3] After the burst: available=" + pool.available()
                    + "/" + pool.size() + ", total rows=" + rows
                    + " (expected " + (TASKS * 10_000L) + ")");
        }

        System.out.println("\n=== Pooling tour complete ===");
    }
}

File: samples/src/main/kotlin/io/github/danielbunting/clickhouse/samples/kotlin/Pooling.kt
package io.github.danielbunting.clickhouse.samples.kotlin

import io.github.danielbunting.clickhouse.samples.ClickHouseEnv
import io.github.danielbunting.clickhouse.kotlin.command
import io.github.danielbunting.clickhouse.kotlin.lease
import io.github.danielbunting.clickhouse.kotlin.openPool
import io.github.danielbunting.clickhouse.kotlin.scalar
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.async
import kotlinx.coroutines.awaitAll
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.runBlocking

/**
 * Concurrency with the connection pool, coroutine-style — the counterpart of the Java
 * [Pooling] sample.
 *
 * A single connection is one socket used by one coroutine at a time; the pool is the unit
 * of concurrency. `openPool` opens the connections off-thread (suspend); `lease { }` is the
 * suspend borrow → work → always-return scope (the return is `NonCancellable`, so a
 * cancelled block can never leak its permit).
 *
 * Steps:
 *  1. Open a pool of 4 connections.
 *  2. Launch 16 coroutines; each leases a connection, inserts 10k rows, and counts its own
 *     partition — 4 run at a time, the rest suspend (not block) waiting for a permit.
 *  3. Show pool health (`available()/size()`) after the burst.
 *
 * Run with:
 * ```
 *   ./gradlew :samples:runKotlinPooling
 * ```
 */
fun main(): Unit = runBlocking {
    val table = "pool_demo_kotlin"
    val poolSize = 4
    val tasks = 16
    println("=== clickhouse-native-client-kotlin Pooling tour ===")

    println("\n[1] Opening a pool of $poolSize connections...")
    ClickHouseEnv.config().openPool(size = poolSize).use { pool ->
        println("    size=${pool.size()}, available=${pool.available()}")

        pool.lease { conn ->
            conn.command(
                "CREATE OR REPLACE TABLE $table (" +
                    "task UInt32," +
                    "n    UInt64" +
                    ") ENGINE = MergeTree ORDER BY (task, n)",
            )
        }

        println("\n[2] Running $tasks concurrent tasks as coroutines over $poolSize connections...")
        val perTask = coroutineScope {
            (0 until tasks).map { task ->
                async(Dispatchers.Default) {
                    pool.lease { conn ->
                        conn.command("INSERT INTO $table SELECT $task, number FROM numbers(10000)")
                        conn.scalar("SELECT count() FROM $table WHERE task = $task")
                    }
                }
            }.awaitAll()
        }
        println("    All tasks done; per-task counts sum to ${perTask.sum()}")

        val rows = pool.lease { conn -> conn.scalar("SELECT count() FROM $table") }
        println("\n[3] After the burst: available=${pool.available()}/${pool.size()}, " +
            "total rows=$rows (expected ${tasks * 10_000L})")
    }

    println("\n=== Kotlin Pooling tour complete ===")
}
